crates/red_knot_python_semantic/resources/mdtest/annotations/annotated.md

```diff
@@ -0,0 +1,79 @@
+# `Annotated`
+
+`Annotated` attaches arbitrary metadata to a given type.
+
+## Usages
+
+`Annotated[T, ...]` is equivalent to `T`: All metadata arguments are simply ignored.
+
+```py
+from typing_extensions import Annotated
+
+def _(x: Annotated[int, "foo"]):
+    reveal_type(x)  # revealed: int
+
+def _(x: Annotated[int, lambda: 0 + 1 * 2 // 3, _(4)]):
+    reveal_type(x)  # revealed: int
+
+def _(x: Annotated[int, "arbitrary", "metadata", "elements", "are", "fine"]):
+    reveal_type(x)  # revealed: int
+
+def _(x: Annotated[tuple[str, int], bytes]):
+    reveal_type(x)  # revealed: tuple[str, int]
+```
+
+## Parameterization
+
+It is invalid to parameterize `Annotated` with less than two arguments.
+
+```py
+from typing_extensions import Annotated
+
+# TODO: This should be an error
+def _(x: Annotated):
```

Comment:
There are two ways to solve this TODO. One way would be to change `Type::in_type_expression` so that it returns a `Result`, where the error tells you what kind of diagnostic needs to be emitted if it's not a valid type expression. The other would be to move `Type::in_type_expression` so that it's a method on `TypeInferenceBuilder` rather than a method on `Type`, since the type inference builder has access to diagnostics. Either would be do the job; they both have advantages and disadvantages. Interested in hearing @carljm's thoughts on what he'd prefer.

Reply:
Moving it to `TypeInferenceBuilder` seems easy, but we currently roughly maintain a split where methods on `TypeInferenceBuilder` accept AST nodes as input, not types; type-level operations go on `Type` and in `types.rs`. I think I'd prefer to maintain that?

I think there's a third option, which might be simplest: have `in_type_expression` take as argument `&self.diagnostics` when called from `TypeInferenceBuilder`. This looks like the direction we might be heading with https://github.com/astral-sh/ruff/pull/14956/files anyway, and would be easy to convert to that.

Reply:
I think this might just be something we have to accept, that type operations can add diagnostics to the current inference context? I'm not sure it's nicer if every call to `in_type_expression` has to look like `some_type.in_type_expression(db).unwrap_with_diagnostic(&mut self.diagnostics)` instead.

Reply:
Yes, fair enough. @InSyncWithFoo, can you make that change, then? `Type::in_type_expression` needs to accept an additional argument, of type `&mut TypeCheckDiagnostics`, so that it can emit diagnostics if the type isn't valid in a type expression

Reply:
Oh, I've forgotten a key point, though -- reporting a diagnostic also requires a node (to report it on the right range). Having `Type::in_type_expression` also take an AST node is just as much a boundary-violation as having methods on `TypeInferenceBuilder` take types. So I feel like returning a `Result` is probably best, or at least most in line with our current structure. Though I'm not certain whether holding this boundary will be sustainable long-term.

Reply:
In this case it feels complicated enough that I'm okay leaving the TODO here for now and accepting the PR as-is, if you are!

Reply:
Yes, I think that's fine. We need to sort out our strategy here, and I'll be working on that as part of call signature checking anyway.